test/edgehog_device_forwarder/websockets_test.exs

```diff
@@ -0,0 +1,134 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.WebSocketsTest do
+  use EdgehogDeviceForwarder.CacheCase
+  use EdgehogDeviceForwarder.MessageCase
+  alias EdgehogDeviceForwarderProto.Edgehog.Device.Forwarder.Message, as: ProtoMessage
+  alias EdgehogDeviceForwarderProto.Edgehog.Device.Forwarder.WebSocket, as: ProtoWebSocket
+  alias EdgehogDeviceForwarder.{HTTPRequests, WebSockets}
+  alias EdgehogDeviceForwarder.WebSockets.Data
+
+  require Logger
+
+  setup_all do
+    [valid_message: {:text, "1"}]
+  end
+
+  setup %{
+    socket: controller,
+    http_upgrade_response: upgrade
+  } do
+    me = self()
+
+    request_id = HTTPRequests.new(controller)
+    socket = spawn(fn -> user_socket() end)
+
+    data = %Data{socket_id: request_id, device: me}
+    :ok = HTTPRequests.forward(request_id, upgrade)
+    :ok = WebSockets.upgrade(request_id, socket, data)
+
+    [
+      socket_id: request_id,
+      device_socket: me,
+      socket_data: data,
+      user_socket: socket
+    ]
+  end
+
+  describe "upgrade/3" do
+    setup %{socket: controller} do
+      request_id = HTTPRequests.new(controller)
+      upgrade_data = %Data{socket_id: request_id, device: self()}
+
+      [http_request_id: request_id, upgrade_data: upgrade_data]
+    end
+
+    test "does not allow an active http request to be upgraded", %{
+      http_request_id: request_id,
+      socket_data: data,
+      socket: websocket_socket
+    } do
+      assert {:error, :http_request_not_found} ==
+               WebSockets.upgrade(request_id, websocket_socket, data)
+    end
+
+    test "saves the new socket to the cache using the old request id", %{
+      http_request_id: request_id,
+      socket_data: data,
+      http_upgrade_response: upgrade
+    } do
+      message = {:text, "test"}
+
+      # We allow upgrading only after forwarding an upgrade request
+      HTTPRequests.forward(request_id, upgrade)
+
+      assert :ok == WebSockets.upgrade(request_id, self(), data)
+      assert :ok == WebSockets.forward(request_id, message)
+      assert_receive ^message
+    end
+  end
+
+  describe "forward/2" do
+    test "sends the message to the target socket", %{
+      socket: controller,
+      socket_data: data,
+      http_upgrade_response: upgrade
+    } do
+      # insert websocket with target self()
+      request_id = HTTPRequests.new(controller)
+      :ok = HTTPRequests.forward(request_id, upgrade)
+      :ok = WebSockets.upgrade(request_id, self(), data)
+
+      # assert that we receive all messages forwarded towards self()
+      messages = [{:text, "1"}, {:text, "2"}]
+
+      for message <- messages do
+        :ok = WebSockets.forward(request_id, message)
+      end
+
+      for message <- messages do
+        assert_receive ^message
```

Comment:
Is the order of messages relevant in this case?

Reply:
right. fixed in 35d6e4d